frappe/workflow/doctype/workflow_action/workflow_action.py

```diff
@@ -3,8 +3,164 @@
-	pass
\ No newline at end of file
+	pass
+
+
+def get_permission_query_conditions(user):
+	if not user: user = frappe.session.user
+
+	if user == "Administrator": return ""
+
+	return "(`tabWorkflow Action`.user='{user}')".format(user=user)
+
+
+def has_permission(doc, user):
+	if user not in ['Administrator', doc.user]:
+		return False
+
+
+def create_workflow_actions(doc, state):
+
+	reference_doctype = doc.doctype
+	reference_name = doc.name
+	workflow = get_workflow_name(reference_doctype)
+
+	if not workflow: return
+
+	workflow = frappe.get_doc('Workflow', workflow)
+
+	doc_current_state = doc.get(workflow.workflow_state_field)
+
+	# hack to check if workflow action was already created and is this method called again due some field changes
+	workflow_action_already_created = frappe.db.count('Workflow Action', {
+		'reference_doctype': reference_doctype,
+		'reference_name': reference_name,
+		'workflow_state': doc_current_state
+	})
+
+	if workflow_action_already_created: return
+
+	immediate_next_possible_transitions = [d for d in workflow.transitions if d.state == doc_current_state]
+
+	current_user = frappe.session.user
+
+	# update status of completed workflow action of user
+	frappe.db.sql('''update `tabWorkflow Action` set status='Completed', completed_by=%s
+		where reference_doctype=%s and reference_name=%s and user=%s''',
+		(current_user, reference_doctype, reference_name, current_user))
+
+	# delete irrelevant workflow actions
+	frappe.db.sql('''delete from `tabWorkflow Action`
+		where reference_doctype=%s and reference_name=%s and user!=%s''',
+		(reference_doctype, reference_name, current_user))
+
+	clear_doctype_notifications('Workflow Action')
+
+	from frappe.permissions import has_permission
+
+	email_map = {}
+
+	for transition in immediate_next_possible_transitions:
+		users = get_users_with_role(transition.allowed)
+		for user in users:
+			if (has_approval_access(doc, workflow, user) and
+				has_permission(doctype=doc, ptype='read', user=user)): pass
+			else: continue
+
+			if not email_map.get(user):
+				email_map[user] = {
+					'possible_actions': [],
+					'email': frappe.db.get_value('User', user, 'email'),
+				}
+
+			email_map[user].get('possible_actions').append({
+				'action_name': transition.action,
+				'action_link': get_workflow_action_url(
+					transition.action,
+					reference_doctype,
+					reference_name,
+					doc_current_state)
+			})
+
+	for user in email_map.keys():
+		newdoc = frappe.get_doc({
+				'doctype': 'Workflow Action',
+				'reference_doctype': reference_doctype,
+				'reference_name': reference_name,
+				'workflow_state': doc_current_state,
+				'status': 'Open',
+				'user': user
+			})
+		newdoc.insert(ignore_permissions=True)
+
+	frappe.db.commit()
+	if email_map:
+		send_workflow_action_email(email_map, reference_doctype, reference_name)
+
+
+def send_workflow_action_email(email_map, doctype, docname):
+	email_message = _('{0}:{1}'.format(doctype, docname))
```

Comment:
make this configurable in State table (?) `next action email template` -> link to `Email Template`